10_rest/willshen8/pkg/puppy/rest.go

```diff
@@ -0,0 +1,118 @@
+package puppy
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/go-chi/chi"
+	"github.com/go-chi/render"
+)
+
+// RestHandler provides a puppy storer to store the puppies
+type RestHandler struct {
+	// sync.Mutex
+	store Storer
+}
+
+// NewRestHandler is a factory method that makes new storer
+func NewRestHandler(storer Storer) *RestHandler {
+	return &RestHandler{store: storer}
+}
+
+// HandleGet gets the puppy by id and displays the results.
+func (rh *RestHandler) HandleGet(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		render.JSON(w, r, &Error{
+			Message: InvalidInput,
+			Code:    HTTPBadRequest,
+		})
+	} else {
+		puppy, readErr := rh.store.ReadPuppy(uint32(id))
+		if readErr != nil {
+			w.WriteHeader(http.StatusNotFound)
+			render.JSON(w, r, &Error{
+				Message: InvalidIDForRead,
+				Code:    HTTPNotFound,
+			})
+		} else {
+			render.JSON(w, r, puppy)
+		}
+	}
+}
+
+// HandlePost post a new puppy and then display the results.
+func (rh *RestHandler) HandlePost(w http.ResponseWriter, r *http.Request) {
+	var p Puppy
+	if decodeErr := render.DecodeJSON(r.Body, &p); decodeErr != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		render.JSON(w, r, &Error{
+			Message: InvalidJSONInput,
+			Code:    HTTPBadRequest,
+		})
+	} else {
+		_, createErr := rh.store.CreatePuppy(&p)
+		if createErr != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			render.JSON(w, r, createErr)
+		} else {
+			render.JSON(w, r, p)
+		}
+	}
+}
+
+// HandlePut updates the existing puppy with new fields in the payload.
+func (rh *RestHandler) HandlePut(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		render.JSON(w, r, &Error{
+			Message: InvliadIDForUpdate,
+			Code:    HTTPBadRequest,
+		})
+	} else {
+		var p Puppy
+		result, updateErr := rh.store.UpdatePuppy(uint32(id), &p)
+		if updateErr != nil {
+			w.WriteHeader(http.StatusNotFound)
```

Comment:
what if I `PUT` puppy with value `-42`? I guess I'll get response code `404` but I should get `422` or `400`

Reply:
I agree with this one. Thanks.